General cleanup, add Win32 support, and add Linux-to-Win32 cross-compilation

rgprog opened the pull request:
This PR adds Win32 support to `capnp.cc`, along with the (extensive!) build infrastructure required to make it available through NPM. See this PR's individual commit messages for the full story.

**As written, this PR will break Rewind**, by (1) changing the CLI commands used for cross-compilation, and (2) refactoring `build.js` and `build-capnp.sh` without testing those changes for any Darwin builds. I can see two options:

- Update Rewind to use the new CLI commands, and then test the various Rewind workflows which produce a Darwin build. Some of the possible failures are quite late-bound in my experience (e.g. failing to dynamically load `capnp.node`), so this would probably require running the resulting builds on an actual Mac.
- Pin Rewind's `capnp` dependency against the most recent commit before this PR (c63332b).

Commit:
Add Win32 support to capnp.cc

This commit crudely edits capnp.cc to support compiling for Windows
as well as Linux, mostly using `#ifdef _WIN32` preprocessor guards.

Thankfully, the underlying libraries (`kj` and `libuv`) are already
platform-agnostic, so the changes could be quite superficial.

The code passes the compiler, but due to linker woes, it hasn't yet
been tested.

Commit:
Refactor build.js, and build libcapnp from scratch on every install

This commit heavily refactors build.js for readability, simplifies our
cross-compilation CLI, introduces static linking for Linux builds, and
sets up the infrastructure for Win32 compilation and cross-compilation
(although both are currently stubbed out).

Commit:
Add native Win32 builds

This commit adds the ability to build the package when both the host
platform and the target platform are `win32`.

This will be soon used to produce a prebuilt `capnp.node`, to facilitate
`linux`-to-`win32` cross-compilation.

Commit:
Statically link the MSVC runtime for Win32 builds

This commit ensures that Win32 builds statically link the C runtime into
`capnp.node`. Its only dynamic dependencies are now `advapi32.dll`,
`kernel32.dll` and `ws2_32.dll`, all of which ship with Windows.

Commit:
Support Linux-to-Win32 cross-compilation, using a prebuilt binary

This commit stores a precompiled Windows binary in
`prebuild/win-x64/capnp.node`, and enables cross-compilation from Linux
by simply copying that binary to the output. (True cross-compilation
isn't possible, because `node-gyp` doesn't support it.)

In the interest of avoiding complication in `git clone`, we store the
binary directly in the repo, rather than using a solution like Git LFS
or an external web host. Each time the binary changes (which should be
infrequent), it will increase the amount of data downloaded in a
`git clone` by about 1.5 MiB.

Commit:
Adjust dynamic binary lookup in capnp.js

Commit:
Add automation of Win32 capnp.node builds

This commit adds a GitHub Actions workflow which can only be run
explicitly. It sets up a native Win32 build environment on a Windows
runner, uses it to produce a prebuilt `capnp.node` for Win32, and then
uploads the result as an artifact.

The purpose of this workflow is to provide a prebuild `capnp.node` for
Linux-to-Win32 cross-compilation, since `node-gyp` doesn't support true
cross-compilation from scratch.

Commit:
Fix build failure

Commit:
Fix Markdown in README.md

Commit:
README.md: Add warning about sticky dependency in NPM

Commit:
Add __dirname to the search path for capnp.importSystem

This is mostly intended as a crude fix to get our `transform/desktop`
app working with `electron-packager`.

Commit:
Call uv_poll_init_socket rather than uv_poll_init

`uv_poll_init` is agnostic between sockets and files on Linux, but not
on Windows. `uv_poll_init_socket` is a socket-specific replacement.

Commit:
Change Windows build settings

`gyp` disables runtime type information (RTTI) by default, and it seems
to have unusual build settings related to C++ exceptions.

`capnp` can tolerate RTTI being disabled, but `node-capnp` can't. As a
workaround, we re-enable RTTI in `binding.gyp`.

Commit:
Make importSystem more permissive, and add it to capnp.d.ts

This commit has `capnp.importSystem` search the current working
directory, as well as `__dirname` and the other standard paths which it
already searched.

Commit:
Improve logging when a Win32 network error occurs

The `kj` library, internal to `capnproto`, has a known bug which causes
it to stringify all Winsock error codes as "Unknown error". This is very
bad for debugging!

In the four places where `KJ_FAIL_SYSCALL` is invoked within `capnp.cc`,
this commit adds additional logging to `stderr`.

Commit:
Fix build failure

Commit:
Print Win32 error codes more nicely

Commit:
Make connect() non-blocking on Win32

Commit:
Update capnp.node to a new version, reflecting recent changes

p--b commented on build-capnp.bat:
```diff
@@ -0,0 +1,46 @@
+:: Syntax: build-capnp.bat
```
As a general principle, I wonder if it would be better for us to be defaulting to powershell scripts rather than batch files - it's a much more pleasant language to work with

rgprog replied:
Moved this one to an offline discussion, for better visibility

p--b commented on src/node-capnp/capnp.cc:
```diff
@@ -47,11 +65,115 @@
+          total = -1;
+        }
+
+        goto end;
```
As `end` doesn't do any cleanup, `return total;` would be clearer here and below.

p--b commented on src/node-capnp/capnp.cc:
```diff
@@ -47,11 +65,115 @@
+    int len = (int) vecs->iov_len;
+    
+    while (len > 0) {
+      int ret = write(fd, buf, len);
```
```suggestion
      int bytes_written = write(fd, buf, len);
```

p--b commented on src/node-capnp/capnp.cc:
```diff
@@ -2097,7 +2274,7 @@ void connect(const v8::FunctionCallbackInfo<v8::Value>& args) {
 }
 
-void connectUnixFd(const v8::FunctionCallbackInfo<v8::Value>& args) {
+/*void connectUnixFd(const v8::FunctionCallbackInfo<v8::Value>& args) {
```
Prefer using `#if 0` rather than commenting out code, ideally with a comment explaining why

p--b commented on src/node-capnp/capnp.cc:
```diff
@@ -2939,7 +3116,7 @@ void init(v8::Local<v8::Object> exports) {
   mapFunction("toBytes", toBytes);
   mapFunction("connect", connect);
-  mapFunction("connectUnixFd", connectUnixFd);
+  //mapFunction("connectUnixFd", connectUnixFd);
```
As above

p--b commented on src/node-capnp/capnp.cc:
```diff
@@ -246,9 +396,9 @@ static constexpr uint NEW_FD_FLAGS =
+  OwnedFileDescriptor(uv_loop_t* loop, SOCKET fd, uint flags)
       : uvLoop(loop), fd(applyFlags(fd, flags)), flags(flags),
-        uvPoller(uv_poll_init, uvLoop, fd) {
+        uvPoller(uv_poll_init_socket, uvLoop, fd) {
```
This seems to be a substantive change from the original that is not win32 only - a comment explaining the reason for the divergence from upstream would be useful, if not putting this in its own commit.

rgprog replied:
Sorry, I'm confused - this is already in its own commit, 80ebee24ce0db73e3b06440d68325e3f7b43259a

(Looks like the commit message is back-to-front! I'll reword it now.)

p--b replied:
Ah, I had missed that it was in its own commit already, apologies. I suppose my point is that given the number commits in this branch currently I was assuming that we would squash it into main, at which point we would lose the reasoning for this change. (TBF We could choose to rebase it in as the commits are all useful, but the PR description doesn't mention that.)

rgprog replied:
The plan is to rebase

p--b commented on build.js:
```diff
@@ -220,27 +364,21 @@ function patchLibs(patchTool, patchPath, target) {
-    ...env
-  };
+if (process.platform === "linux" && args.targetPlatform === "win32") {
+  moveBuildResult('prebuilt/win32-x64/capnp.node', buildEnvironment);
```
Logging that we're reusing a precompiled binary might be a helpful pointer if someone is debugging things in future.

p--b commented on README.md:
```diff
@@ -1,243 +1,64 @@
-# Cap'n Proto bindings for Node.js
+# `node-capnp`
```
Moving the readmes around like this is likely to cause much faff when attempting to merge changes from upstream, hence why our own readme was previously added  as a separate file. By all means we can add a pointer to our readme to the upstream readme, but I don't think we should move them around like this.

rgprog replied:
It took several months for me to realise that `README-FOURIER.md` existed at all, which is why I leaned hard towards maximum visibility here - I'd expect our readme to be read much more often than it's edited.

That being said, it looks like Git's rename detection has already failed us here, so editing a link into the original readme does sound like a pretty fair compromise.

p--b commented on package.json:
```diff
@@ -22,11 +22,11 @@
   "scripts": {
     "install": "node ./build.js --dist-url=https://electronjs.org/headers --target=20.3.3",
-    "test": "electron src/node-capnp/capnp-test.js",
-    "build:darwin": "npm run install -- --target_plat=darwin --patch_path=.libs"
```
I don't see anything that causes the new toolchain to be run with the replacement for `patch_path` (`patch-lib-path`) which will break the darwin build.

rgprog replied:
Good catch - solved this by hardcoding the `patchLibPath`, since the build paths themselves are currently hardcoded anyway.

p--b reviewed (approved):
Given that we're planning on shipping a Darwin client in the first instance, I think we should ensure that these changes result in a functional Darwin module before merging.

Separately, for Rewind, I'm happy for us to pin to an older commit for now rather than adapting its toolchain for this new version; a PR doing that should occur before merging this PR.

p--b commented on src/node-capnp/capnp.cc:
```diff
@@ -47,11 +65,115 @@
+    size_t       iov_len;
+};
+
+int writev(SOCKET fd, const struct iovec* vecs, int count) {
```
As per offline discussion.

p--b reviewed (changes requested):
As per offline discussion

Commit:
Comment out connectUnixFd more nicely

Commit:
Improve logging when using a prebuilt binary

Commit:
Remove --patch-lib-path argument from build.js

We replaced this argument with a hardcoded value, since the build paths
themselves are currently hardcoded. This indirectly fixes a bug where
`--patch-lib-path` was never actually being passed to `build.js` by any
of the run-scripts in `package.json`.

Commit:
Revert README-ORIGINAL.md back to README.md

Due to an increased risk of upstream merge conflicts, the
Fourier-specific `README.md` seems unworkable. Instead, to increase the
visibility of `README-FOURIER.md`, we add a link to the top of the
original `README.md`.

Commit:
Remove use of writev

`writev` is not available on Windows, and its only use in `capnp.cc`
can be straightforwardly replaced with a `for` loop.

rgprog commented:
> Given that we're planning on shipping a Darwin client in the first instance, I think we should ensure that these changes result in a functional Darwin module before merging.

No worries - what's our plan for testing that? I have no access to a Mac.

Commit:
Add support for cross-compilation between architectures

We currently support a `--platform=[win32|darwin|linux]` flag, passed to
any `npm` invocation, to cross-compile to the specified platform.

This commit adds very similar support for an `--arch=[ia32,x64,arm]`
flag, when cross-compiling between architectures.

Commit:
Switch to C++17 from C++14

Recent versions of `v8` require C++17. This commit changes any
command-line requests for C++14 so that they request C++17 instead.

p--b commented:
> what's our plan for testing that

Henry should be able to run the skeletal client that we've built on his laptop - it won't have anything to connect to, but even getting a connection failed error message is a proof of life

Commit:
Change patch_tool version to 21.2

Commit:
Change osxcross SDK to 11.3

Commit:
Change dylib version to 0.10.3

p--b committed:
Remove added import paths breaking ASAR-encapsulated binaries

p--b commented:
Unfortunately this branch also currently results in the client hanging on Linux, seeming to jam the event loop after a request is dispatched on a capability. I have narrowed this down to the source changes to `capnp.cc` (i.e. it doesn't seem to be caused by the build system changes) - checking out `capnp.cc` from main and rebuilding produces a functioning module.

Update: I thought the culprit is the change `uv_poll_init -> uv_poll_init_socket`, but I was mistaken. It seems to be caused by the code to avoid writev...

p--b commented on src/node-capnp/capnp.cc:
```diff
@@ -463,31 +586,47 @@ class UvIoStream: public OwnedFileDescriptor, public kj::AsyncIoStream {
+          fd,
+          (const void*) (((const byte*) thisIov->iov_base) + bytesWritten),
+          thisIov->iov_len - bytesWritten
+        )) {
+          // Error.
+
+          // We can't "return kj::READY_NOW;" inside this block because it causes a memory leak 
+          // due to a bug that exists in both Clang and GCC:
+          //   http://gcc.gnu.org/bugzilla/show_bug.cgi?id=33799
+          //   http://llvm.org/bugs/show_bug.cgi?id=12286
+          goto error;
+        }
+        if (false) {
+        error:
+          return kj::READY_NOW;
+        }
 
-      // We can't "return kj::READY_NOW;" inside this block because it causes a memory leak due to
-      // a bug that exists in both Clang and GCC:
-      //   http://gcc.gnu.org/bugzilla/show_bug.cgi?id=33799
-      //   http://llvm.org/bugs/show_bug.cgi?id=12286
-      goto error;
-    }
-    if (false) {
-    error:
-      return kj::READY_NOW;
+        // A negative result means EAGAIN, which we can treat the same as having written zero bytes.
+        size_t n = writeResult < 0 ? 0 : writeResult;
+        
+        bytesWritten += n;
+        totalBytesWritten += n;
+      }
     }
 
-    // A negative result means EAGAIN, which we can treat the same as having written zero bytes.
-    size_t n = writeResult < 0 ? 0 : writeResult;
+    size_t n = totalBytesWritten;
```
This change seems to cause Linux (and potentially macos) clients to hang their event loop as soon as a capability is used to dispatch an RPC request.

p--b replied:
(Applying [this commit](https://github.com/fourieraudio/node-capnp/commit/d6580a7009e39756a247b3a8270311c909a79e6c), reverting the change, allows the client to run normally on Linux)

rgprog replied:
@p--b: On the current `win32-cross` branch, would you expect this bug to reproduce if I `cargo run --release` Director, start up the new Transform frontend (the `client-infra` branch of Transform), and connect to `127.0.0.1`?

If so, this bug isn't reproducing on Windows. I don't have a Linux test-case which I can use to investigate that side of it (because our frontend doesn't run on WSL, seemingly due to filesystem permission errors). I could try to fix this by staring at it and thinking hard, but the first five minutes of that haven't led to any epiphanies, except that the `writeResult < 0` comparison looks a bit suspicious.

Do you want to look into this, or should we simply `#ifdef` it and pray for forgiveness later?

p--b replied:
Yes, that ought to be enough. I must confess I also can't see what's wrong by inspection. At this stage I think let's go ifdef; writev is possibly marginally preferable anyway for perf.

rgprog replied:
This should now be patched

Commit:
Reintroduce writev on Linux

An undiagnosed bug is preventing our writev replacement from functioning
on Linux. This commit uses #ifdef to reintroduce the old, working code
when not compiling for Win32.

rgprog commented:
Unless there's any urgency, I'd like to hold off on merging this PR until I've made some progress on TF-361 and TF-362; I have a feeling that both tickets are likely to expose additional bugs.

henryharrod commented:
Any update on this PR? Got mine dependent on it (in order to build transform desktop)

rgprog commented:
TF-362 was finished (pending review) without exposing any new bugs. TF-361 is still pending, but perhaps a little less relevant to this PR, because it should bear a pretty close resemblance to our existing frontend capnproto code.

No objections to merging now (by rebase), if it would make your life easier! It's just blocked on an approving code review from @p--b.

p--b commented on src/node-capnp/capnp.cc:
```diff
@@ -463,18 +586,55 @@ class UvIoStream: public OwnedFileDescriptor, public kj::AsyncIoStream {
                                   kj::ArrayPtr<const kj::ArrayPtr<const byte>> morePieces) {
     KJ_STACK_ARRAY(struct iovec, iov, 1 + morePieces.size(), 16, 128);
 
-    // writev() interface is not const-correct.  :(
```
nitpick: given that we might upstream this, probably worth not removing this

p--b commented on build.js:
```diff
@@ -22,153 +22,291 @@
+} else if (process.platform === "win32") {
+  capnpLibPath = path.resolve("./build-capnp/.libs");
+  capnpIncPath = path.resolve("./build-capnp/src/");
+  patchLibPath = undefined; // Unused, since we don't support win32->darwin builds
```
Is it worth tagging on an `else { throw new Error("Unsupported platform") }` in case we try to add building from `darwin` in the future (I know it's tested before above but makes the change easier to make in future)

p--b commented on build.js:
```diff
@@ -182,65 +320,71 @@ function afterBuild() {
  * @param patchPath - the path to be stripped out of the binary
  * @param target - The target binary to be modified
  */
+
```
```suggestion
```

p--b commented on build.js:
```diff
@@ -182,65 +320,71 @@ function afterBuild() {
-  }).concat([target]);
-
-  console.log(`Invoking binary patch tool ${patchTool} with the following arguments:`);
+  var libs = ["libkj", "libkj-async", "libcapnp", "libcapnpc", "libcapnp-rpc"];
```
Nitpick: changes like this one (all diff :327->:357 of `patchLibs`) don't seem to add anything to the functionality but do clutter the history & increase the size of the diff. Unless this change was made automatically by a formatter, consider avoiding changes like this to make it easier to trace the history of the code & reduce review size - it's really not clear from this diff that the only substantive change to this function is on :325.

p--b commented on binding.gyp:
```diff
@@ -12,15 +11,52 @@
+
+          # Static linking
+          'libraries': [
+            '-Wl,--whole-archive',
```
Why are we adding `--whole-archive`? Deserves a comment

p--b replied:
I know what it does, I don't know _why_ we're using it here: I don't see why we should be shipping bits of libkj we're not using

rgprog replied:
I believe introducing the flag resolved some stubborn "undefined symbol" errors, but it's been nearly two months, so I can't recall the fine details. I don't think I investigated whether the flag could be omitted for individual static libraries like `libkj`.

Properly reevaluating whether the `--whole-archive` flag is necessary here would take experimentation. If you'd like me to spend a bit of time on that, let me know.

p--b commented on binding.gyp:
```diff
@@ -12,15 +11,52 @@
+            '../build-capnp/.libs/libcapnp.a',
+            '../build-capnp/.libs/libcapnpc.a',
+            '../build-capnp/.libs/libcapnp-rpc.a',
+            '-Wl,--no-whole-archive',
```
This seems odd, given :34. If it's correct, deserves a comment

p--b commented on build-capnp.sh:
```diff
@@ -0,0 +1,88 @@
+if [[ $1 == "linux" ]]; then
+    # We're not cross-compiling, so this is trivial.
+
+    CC="clang" \
+        CXX="clang++" \
+        CFLAGS="-fPIC" \
+        CXXFLAGS="-std=c++17 -fPIC" \
+        ./configure --disable-shared
+    make -j
+    make install-data DESTDIR=capnp-root
```
Nitpick: this is a substantive change: previously on linux we linked against the system's libcapnp. Whilst I'm not against this in principle for reproducibility reasons, I would have expected this to have been called out more visibly in this PR.

p--b commented on build-capnp.sh:
```diff
@@ -0,0 +1,88 @@
+
+# Download and unpack the capnproto source. It's pretty small (1.6 MiB), so this should be fast.
+curl -O https://capnproto.org/capnproto-c++-0.10.3.tar.gz
+tar --strip-components=1 -zxf capnproto-c++-0.10.3.tar.gz
```
Please check if this already exists and reuse it if it does. Redownloading stuff every time we do a build is wasteful, adds up fast, and makes developing offline / remotely difficult.

p--b commented on build-capnp.sh:
```diff
@@ -0,0 +1,88 @@
+pushd ${build_dir}
+
+# Download and unpack the capnproto source. It's pretty small (1.6 MiB), so this should be fast.
+curl -O https://capnproto.org/capnproto-c++-0.10.3.tar.gz
```
Now that we are not using the system's libcapnp, `install-deps.sh` in this repository is unnecessary and should be removed.

henryharrod commented:
That would be great - thanks @rgprog. It's currently holding up TF-13 as that depends on this main branch (node-capnp) being up to speed with the changes in win32-cross-revert-writev (which is itself dependent on this PR merging). 

If you could do this week that would be great, as then can try to merge TF-13 by the time I go away next Weds.

Ta1

p--b reviewed (approved):
Looks good pending comments.

As a nitpick, this PR has a lot of commits which aren't really the sort of commits I'd want to see in the history (e.g. 5b0a55dc, 41941fa9, dc0c6e61, 85005449, f7f699f, 1e74538). If we were squashing this branch it would be irrelevant, but given we plan to rebase, having commits like those in the history make it much more difficult to helpfully use `git log` / `git blame` to understand when, how (and critically _why_) changes were made to the software.

In the interests of time we might just want to merge this (as the time we've spent on enabling Windows development is already significantly above what it would have taken to just run a VM, and this work is not user visible in the short term), but in general (and especially for work that we might consider upstreaming), our future selves will thank us for not making a mess of the repository history.

Commit:
Changes suggested during code review

Commit:
Update prebuilt/win32-x64/capnp.node to reflect recent changes

Commit:
Adjust CI to reflect recent changes

Commit:
Revert "Adjust CI to reflect recent changes"

This reverts commit 2c4f3584d80ee3ed6c11ad73f9ea162b96362dec.

Commit:
Reintroduce `install-deps.sh`, as it's required to run `npm test`

Commit:
Change permissions on install-deps.sh
